### ISSUE TITLE
About training omnidirectional walking

### ISSUE THREAD

**Issue by zwww-www** [m-abr/FCPCodebase#42]
In the open source code, there is already an omnidirectional walking model, I do not know whether its walking speed can continue to improve, so I want to train again.When reading the paper, I did not understand the content in table12, so thank you for explaining it to me.

![Image](https://github.com/user-attachments/assets/2b9d816f-05b8-4a53-8499-7c1cf1358473)

**Comment by m-abr** [comment]
Each episode involves two targets: a **target position (TP)** and a **target direction (TD)**. The TP moves within a square work area measuring 14 m × 14 m = 196 m², with bounds from -7 m to 7 m along both the x and y axes. A random 2D velocity is periodically assigned to TP, on average every 1.6 seconds, causing it to move continuously within this area. If TP exits the work area, it is immediately relocated to the center of the field (i.e., at coordinates (0, 0)).

The target direction TD is handled differently depending on the episode:

- **In 30% of episodes**, TD is aligned with TP, meaning the robot is encouraged to move directly toward the target — promoting forward walking, which is the most common locomotion type in actual matches.

- **In the remaining 70% of episodes**, TD changes independently of TP at random intervals (on average every 4 seconds). This setup ensures the robot also learns **omnidirectional walking**, crucial for tasks like aligning for a kick or repositioning efficiently during gameplay.

---

### Reward Function

`r = progress x direction_penalty + idle_bonus`

- **progress**: the reduction in distance to TP since the last visual update. It encourages the robot to move closer to the target.

- **direction_penalty**: defined as `1.03^(-direction_error)`, it penalizes deviations in heading direction, where `direction_error` is the angular difference (in degrees) between the robot’s orientation and TD.

- **idle_bonus**: a small positive reward awarded when the robot is within 0.2 m of TP. This bonus encourages both proximity to the target and minimal movement — i.e., low-magnitude action outputs from the neural network — allowing the locomotion system to revert to its underlying stepping behavior. This incentivizes approaching the target with precision and stability.

**Comment by zwww-www** [comment]
> 每个事件涉及两个目标：**目标位置 （TP）** 和**目标方向 （TD）。**TP 在 14 m × 14 m = 196 m² 的方形工作区域内移动，沿 x 轴和 y 轴的边界为 -7 m 至 7 m。TP 会定期分配一个随机的 2D 速度，平均每 1.6 秒一次，使其在该区域内连续移动。如果 TP 退出工作区域，它会立即重新定位到场的中心（即坐标 （0， 0））。
> 
> 目标方向 TD 的处理方式因剧集而异：
> 
> * **在 30% 的情节中**，TD 与 TP 对齐，这意味着鼓励机器人直接向目标移动——促进向前行走，这是实际比赛中最常见的运动类型。
> * **在其余 70% 的发作中**，TD 以随机间隔（平均每 4 秒）独立于 TP 变化。这种设置可确保机器人还学习**全向行走**，这对于在游戏过程中有效地对齐踢腿或重新定位等任务至关重要。
> 
> ### 奖励功能
> `r = progress x direction_penalty + idle_bonus`
> 
> * **progress**：自上次视觉更新以来到 TP 的距离减少。它鼓励机器人更接近目标。
> * **direction_penalty**：定义为 ，它惩罚前进方向的偏差，其中 是机器人方向与 TD 之间的角度差（以度为单位）。`1.03^(-direction_error)``direction_error`
> * **idle_bonus**：当机器人在 TP 的 0.2 m 范围内时，会获得一个小的正奖励。这种好处既鼓励接近目标，又鼓励最小运动（即神经网络的低幅度动作输出），从而允许运动系统恢复到其基本的步进行为。这激励了精确和稳定地接近目标。

Does the reward function update every 0.04s or only the progress function update every 0.04s?

**Comment by zwww-www** [comment]
Thank you very much for your help. I have completed the code, but I don't know if it meets the overall requirements of the paper. It gets stuck after running to 600,000 time steps. I don't know what's going on (as shown below). Finally, I will send the code for your reference.

![Image](https://github.com/user-attachments/assets/f11959c5-82fb-4198-ba06-ff4f25f11c13)

[Walk.txt](https://github.com/user-attachments/files/19645170/Walk.txt)

**Comment by m-abr** [comment]
Regarding the reward function: the full reward is only computed during vision cycles (when `player.world.vision_is_up_to_date` is `True`).

I haven’t had time to go through your entire code, but based on your description, I think the issue you're encountering (where the simulation gets stuck) is due to how evaluation episodes are handled. According to your code, evaluation is triggered every 20 policy update iterations (`eval_freq=n_steps_per_env*20`), and during each evaluation stage, the agent must complete 5 episodes (by default) to compute average reward and episode length.

The problem arises if the robot has learned to remain stationary without falling. Since your current terminal condition only ends the episode if the robot falls (i.e., `terminal = r.cheat_abs_pos[2] < 0.35`), then during evaluation, episodes can go on indefinitely if the robot simply stays upright. This causes the evaluation stage to hang.

To confirm this, you can monitor the evaluation environment on port `4216` (since you're running 16 threads with monitor ports from `4200` to `4215`). You’ll likely observe the agent making slight movements while upright.

To fix this, I recommend adding a time-limit to episodes during evaluation. You can do this by modifying your terminal condition like so:

```python
terminal = r.cheat_abs_pos[2] < 0.35 or (self.step_counter > 500 and self.is_evaluation)
```

This limits each evaluation episode to a maximum of 500 steps (which is equivalent to 10 seconds of simulated time). You can adjust the `500` to any value you prefer.

To support this change, you’ll also need to define `self.is_evaluation`. Update your environment class constructor like this:

```python
def __init__(self, ip, server_p, monitor_p, r_type, enable_draw, is_evaluation=False) -> None:
    ...
    self.is_evaluation = is_evaluation
```

And update the `init_env` function accordingly:

**Before:**
```python
def init_env(i_env):
    def thunk():                                                                               # False
        env = Walk(self.ip, self.server_p + i_env, self.monitor_p_1000 + i_env, self.robot_type, True)
```

**After:**
```python
def init_env(i_env, is_evaluation=False):
    def thunk():                                                                               # False
        env = Walk(self.ip, self.server_p + i_env, self.monitor_p_1000 + i_env, self.robot_type, True, is_evaluation)
```

Finally, make sure to initialize your evaluation environments like this:

**Before:**
```python
eval_env = SubprocVecEnv([init_env(n_envs)])
```

**After:**
```python
eval_env = SubprocVecEnv([init_env(n_envs, True)])
```

Hope this helps!

**Comment by qweasd982** [comment]
Thank you very much for your help. As you said, during the training of omnidirectional walking, the model moves slowly and does not fall down. I will modify the code according to your suggestions and test again.

**Comment by qweasd982** [comment]
[Walk.txt](https://github.com/user-attachments/files/19676792/Walk.txt)   Modified version

**Comment by zwww-www** [comment]
Thank you very much for your help. I have modified the code according to your suggestions. I will feedback the training effect of the code after testing, and attach the modified code at last.

[Walk.txt](https://github.com/user-attachments/files/19676894/Walk.txt)

**Comment by zwww-www** [comment]
After training 32 million time steps, the test was performed, and the results, as shown in the video and images below, were not good. I think he fits the description of the paper. I don't know what the problem is.

![Image](https://github.com/user-attachments/assets/03a48259-5bca-438d-86b5-5edcb9c0e1b6)
https://github.com/user-attachments/assets/40215e6e-9491-410b-8f2f-32af08f4581a

**Comment by zwww-www** [comment]
In the paper, TD and TP are updated every Episode. I just limited the number of steps in the test environment. If the agent always moves slowly and does not fall during training, TD and TP will never be updated again. Will this be the reason for the poor training effect?

**Comment by m-abr** [comment]
I’m not sure why the performance is poor. I haven’t had time to thoroughly look into the gym file you shared. Regarding TD and TP: in your code, it looks like they’re being updated continuously during the episode, which is also how it's handled in my paper. I mentioned earlier that TD and TP are both assigned periodically within each episode (not only at the start of each episode), so I’m not sure what you're referring to when you say they’re only updated per episode in the paper

**Comment by zwww-www** [comment]
I'm very sorry, I didn't explain it clearly enough. My understanding is as follows: at the beginning of each episode, in the reset() method, we randomly assign TD to be along the direction of TP, or change it randomly independently of TP. If in reset(), TD is along the direction of TP, it will continue to be along the direction of TP until the agent falls and reset() is executed again. In the code:
`# reset()
        if np.random.rand() <= 0.3:
            self.update_direction = False
            self.target_direction = np.degrees(np.arctan2(self.walk_rel_target[1]-r.cheat_abs_pos[1], self.walk_rel_target[0]-r.cheat_abs_pos[0]))
        else:
            self.update_direction = True
            self.target_direction = np.random.uniform(-180, 180)`
where self.update_direction represents whether TD is along the TP direction or randomly assigned in this episode.

**Comment by zwww-www** [comment]
I'm very sorry, I didn't explain it clearly enough. My understanding is as follows: at the beginning of each episode, in the reset() method, we randomly assign TD to be along the direction of TP, or change it randomly independently of TP. If in reset(), TD is along the direction of TP, it will continue to be along the direction of TP until the agent falls and reset() is executed again. In the code:
```
# reset()
        if np.random.rand() <= 0.3:
            self.update_direction = False
            self.target_direction = np.degrees(np.arctan2(self.walk_rel_target[1]-r.cheat_abs_pos[1], self.walk_rel_target[0]-r.cheat_abs_pos[0]))
        else:
            self.update_direction = True
            self.target_direction = np.random.uniform(-180, 180)
```
where self.update_direction represents whether TD is along the TP direction or randomly assigned in this episode.

**Comment by zwww-www** [comment]
I'm very sorry, I didn't figure out whether TP and TD are relative or absolute. Below is the latest code.

[Walk.txt](https://github.com/user-attachments/files/19719698/Walk.txt)

**Comment by zwww-www** [comment]
Again, I made some changes to the code, which can be trained at present, but there may be some problems, due to resource constraints, I only trained 37 million time steps. I will give the code, I hope you can help me to correct the mistakes and shortcomings. The current training results can be seen by clicking on the link.

[Walk.txt](https://github.com/user-attachments/files/19725545/Walk.txt)

https://www.bilibili.com/video/BV1JAd6YbEkA/?spm_id_from=333.1387.upload.video_card.click

**Comment by zwww-www** [comment]
I want to know more about omnidirectional walking. Do you have any papers on it?

**Comment by m-abr** [comment]
The paper I have on omnidirectional walking is the one related to this codebase ([https://doi.org/10.1007/s00521-025-11151-3](https://doi.org/10.1007/s00521-025-11151-3)). Unfortunately, I don’t have time to review the attached file at the moment, but I suggest extending the training duration to see if the skill improves. Developing an effective omnidirectional walk does require significant effort, but it’s definitely achievable with persistence

**Comment by zwww-www** [comment]
I chose type 1 robot and used the above code to train for 50 million time steps. I tested the robot's walking speed, which was only half of the speed in the open source model. There may be some problems with the code. I hope you can help me correct them. Thank you very much.
